modyn/evaluator/internal/pytorch_evaluator.py

```diff
@@ -0,0 +1,177 @@
+        return self._mod_dict[func_name]
+
+    def _info(self, msg: str) -> None:
+        self.logger.info(f"[Evaluation {self._evaluation_id}] {msg}")
+
+    def _debug(self, msg: str) -> None:
+        self.logger.debug(f"[Evaluation {self._evaluation_id}] {msg}")
+
+    def _load_state(self, path: pathlib.Path) -> None:
+        assert path.exists(), "Cannot load state from non-existing file"
+
+        self._info(f"Loading model state from {path}")
+        with open(path, "rb") as state_file:
+            checkpoint = torch.load(io.BytesIO(state_file.read()))
+
+        assert "model" in checkpoint
+        self._model.model.load_state_dict(checkpoint["model"])
+
+        # delete trained model from disk
+        os.remove(path)
+
+    def send_status_to_server(self, batch_number: int) -> None:
+        self._status_response_queue.put({"num_batches": batch_number, "num_samples": self._num_samples})
+
+    def evaluate(self) -> None:
+        self._info(f"Process {os.getpid()} starts evaluation.")
+
+        self._model.model.eval()
+        with torch.no_grad():
+            batch_number = -1
+            for batch_number, batch in enumerate(self._dataloader):
+                # As empty() is unreliable
+                # we try to fetch an element within 10ms. If there is no
+                # element within that timeframe returned, we continue.
+                try:
+                    req = self._status_query_queue.get(timeout=0.01)
+                    if req == EvaluatorMessages.STATUS_QUERY_MESSAGE:
+                        self.send_status_to_server(batch_number)
+                    else:
+                        raise ValueError("Unknown message in the status query queue")
+                except queue.Empty:
+                    pass
+
+                data: Union[torch.Tensor, dict]
+                if isinstance(batch[1], torch.Tensor):
+                    data = batch[1].to(self._device)
+                elif isinstance(batch[1], dict):
+                    data: dict[str, torch.Tensor] = {}  # type: ignore[no-redef]
+                    for name, tensor in batch[1].items():
+                        data[name] = tensor.to(self._device)
+
+                if self._label_tranformer_function is None:
+                    target = batch[2].to(self._device)
+                else:
+                    target = self._label_tranformer_function(batch[2]).to(self._device)
+
+                batch_size = target.shape[0]
+
+                with torch.autocast(self._device_type, enabled=self._amp):
+                    output = self._model.model(data)
+
+                    if self._evaluation_transformer:
+                        output = self._evaluation_transformer(output)
+
+                    for metric in self._metrics:
+                        metric.evaluate_batch(target, output, batch_size)
+
+                self._num_samples += batch_size
+
```

Comment:
can you add a todo here (and create an issue) to support holistic metrics? I think it should be the goal of your thesis that we support at least F1, Accuracy, and AUC :)

Reply:
Yes, sure :)
I didn't add a TODO here as I want to implement the functionality in this PR.